Parte 2 -   Desarrollo   de   un   modelo   de   Clasificación   para   determinar   si   el   tumor   en
cuestión es benigno o malignp

In [1]:
!pip install scikit-learn

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [2]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import numpy as np
import pandas as pd

In [39]:
data = pd.read_csv('cleaned_data.csv')

In [6]:
#para que solo hayan dos clases de respuesta se editara la columna class, donde 2 sera 0 (benigno) y 4 sera 1 (maligno)
data['class'] = data['class'].replace({2: 0, 4: 1})

In [7]:
#este paso solo es para confirmar
data

id  clump_thickness  uniformity_of_cell_size  \
0    1000025                5                        1   
1    1002945                5                        4   
2    1015425                3                        1   
3    1016277                6                        8   
4    1017023                4                        1   
..       ...              ...                      ...   
486   654546                1                        1   
487   695091                5                       10   
488   714039                3                        1   
489   776715                3                        1   
490   841769                2                        1   

     uniformity_of_cell_shape  marginal_adhesion  single_epithelial_cell_size  \
0                           1                  1                            2   
1                           4                  5                            7   
2                           1                  1                            2   
3                           8                  1                            3   
4                           1                  3                            2   
..                        ...                ...                          ...   
486                         1                  3                            2   
487                        10                  5                            4   
488                         1                  1                            2   
489                         1                  1                            3   
490                         1                  1                            2   

    bare_nuclei  bland_chromatin  normal_nucleoli  mitoses  class  
0             1                3                1        1      0  
1            10                3                2        1      0  
2             2                3                1        1      0  
3             4                3                7        1      0  
4             1                3                1        1      0  
..          ...              ...              ...      ...    ...  
486           1                1                1        1      0  
487           5                4                4        1      1  
488           1                1                1        1      0  
489           2                1                1        1      0  
490           1                1                1        1      0  

[491 rows x 11 columns]

In [41]:
data['bare_nuclei'] = pd.to_numeric(data['bare_nuclei'], errors='coerce')
data['bare_nuclei'] = data['bare_nuclei'].astype('Int64')
data = data.dropna()


In [43]:
data.drop('id', axis=1, inplace=True)

# Regresion Logistica

In [46]:
X = data.drop('class', axis=1)
Y = data['class']

X_entreno, X_prueba, Y_entreno, Y_prueba = train_test_split(X, Y, test_size=0.5, random_state=42)

LR = LogisticRegression()
LR.fit(X_entreno, Y_entreno)

LogisticRegression()

In [47]:
Y_prediccion = LR.predict(X_prueba)
precision = accuracy_score(Y_prueba, Y_prediccion)
confusion = confusion_matrix(Y_prueba, Y_prediccion)
clasificacion = classification_report(Y_prueba, Y_prediccion)

In [48]:
print(f'Precisión del modelo: {precision}')

Precisión del modelo: 0.9708333333333333


In [49]:
print(f'\Matriz de confusión del modelo :\n{confusion}')

\Matriz de confusión del modelo :
[[206   2]
 [  5  27]]


In [50]:
print(f'\Reporte de clasificación del modelo:\n{clasificacion}')

\Reporte de clasificación del modelo:
              precision    recall  f1-score   support

           2       0.98      0.99      0.98       208
           4       0.93      0.84      0.89        32

    accuracy                           0.97       240
   macro avg       0.95      0.92      0.93       240
weighted avg       0.97      0.97      0.97       240



# KNN

In [52]:
from sklearn.neighbors import KNeighborsClassifier

In [53]:
X_entrenoknn, X_pruebaknn, Y_entrenoknn, Y_pruebaknn = train_test_split(X, Y, test_size=0.5, random_state=42)

KNN = KNeighborsClassifier(n_neighbors=6)
KNN.fit(X_entrenoknn, Y_pruebaknn)

KNeighborsClassifier(n_neighbors=6)

In [54]:
y_prediccionknn = KNN.predict(X_pruebaknn)
precicionknn = accuracy_score(Y_pruebaknn, y_prediccionknn)
confusionknn = confusion_matrix(Y_pruebaknn, y_prediccionknn)
classificacionknn = classification_report(Y_pruebaknn, y_prediccionknn)


In [55]:
print(f'Precisión del modelo: {precicionknn}')

Precisión del modelo: 0.8333333333333334


In [56]:
print(f'Matriz de confusión del modelo:\n{confusionknn}')

Matriz de confusión del modelo:
[[200   8]
 [ 32   0]]


In [57]:
print(f'\Reporte de clasificación:\n{classificacionknn}')

\Reporte de clasificación:
              precision    recall  f1-score   support

           2       0.86      0.96      0.91       208
           4       0.00      0.00      0.00        32

    accuracy                           0.83       240
   macro avg       0.43      0.48      0.45       240
weighted avg       0.75      0.83      0.79       240



# CONCLUSION

Como se puede observar en las precisiones, la de la regresion fue de 0.97 y la de KNN de 0.83. Por lo tanto se puede decir que la regresion logistica es mas precisa y acertiva en cuanto a los resultados de las clases de cáncer de la mama